src/gwt/panmirror/src/editor/src/optional/ace/render_queue.ts

```diff
@@ -0,0 +1,215 @@
+/*
+ * render_queue.ts
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+import { CodeBlockNodeView } from './ace';
+
+/**
+ * Represents a queue of Ace editor instances that are rendered asynchronously.
+ */
+export class AceRenderQueue {
+  private readonly renderQueue: CodeBlockNodeView[] = [];
+
+  private renderCompleted: boolean = false;
+  private renderTimer: number = 0;
+  private container?: HTMLElement;
+  private needsSort: boolean = true;
+  private bypass: number = 5;
+  private observer?: IntersectionObserver;
+  private visible: boolean = true;
+
+  /**
+   * Sets (or replaces) the scroll container hosting the render queue. The
+   * scroll container is used to prioritize the queue (i.e. objects in the
+   * viewport or close to it are to be given more priority).
+   * 
+   * @param container The HTML element of the scroll container
+   */
+  public setContainer(container: HTMLElement) {
+
+    // Handler for scroll events in the container
+    const handleScroll = (evt: Event) => {
+      // Whenever a scroll event occurs, we need to re-sort the queue so that
+      // objects in or closest to the viewport are given priority.
+      this.needsSort = true;
+
+      // If we don't think we're visible but we're scrolling and have height,
+      // then we are in fact visible. (This catches a case where the
+      // intsersection observer created below doesn't fire for visiblity
+      // changes.)
+      if (!this.visible && this.container && this.container.offsetHeight > 0) {
+        this.visible = true;
+        this.processRenderQueue();
+      }
+    };
+
+    // Skip if we're already looking at this container
+    if (this.container === container) {
+      return;
+    }
+
+    // Clean up handlers on any previous container
+    if (this.container) {
+      this.container.removeEventListener("scroll", handleScroll);
+    }
+    if (this.observer) {
+      this.observer.disconnect();
+    }
+
+    this.container = container;
+
+    // Create intersection observer to ensure that we don't needlessly churn
+    // through renders while offscreen.
+    this.observer = new IntersectionObserver((entries: IntersectionObserverEntry[]) => {
```

Comment:
Do we need to support IE11?

Reply:
That's a good question. @jjallaire do you know whether visual mode is supported in IE11? (IntersectionObserver has a pretty straightforward polyfill if we need one)

Reply:
I believe that we've tried to preserve support for IE11. This is driven by IDE requirements though, so if the IDE decides to drop IE11 visual mode is fine dropping it. Note that we added this polyfill to the IDE already for IE11: https://github.com/rstudio/rstudio/tree/master/src/gwt/www/js/core-js. So your call either drop IE11 or add the polyfill.